tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.hpp

```diff
@@ -26,32 +26,64 @@ namespace SPH
-			strain_tensor_3D_[index_i] += strain_rate_3D_[index_i] * dt * 0.5;
-			elastic_strain_tensor_3D_[index_i] += elastic_strain_rate_3D_[index_i] * dt * 0.5;
+		}
+		//=================================================================================================//
+		template <class RiemannSolverType>
+		void BaseStressRelaxation1stHalf<RiemannSolverType>::interaction(size_t index_i, Real dt)
+		{
+			Matd velocity_gradient = Matd::Zero();
+			const Neighborhood& inner_neighborhood = inner_configuration_[index_i];
+			for (size_t n = 0; n != inner_neighborhood.current_size_; ++n)
+			{
+				size_t index_j = inner_neighborhood.j_[n];
+				Vecd nablaW_ijV_j = inner_neighborhood.dW_ijV_j_[n] * inner_neighborhood.e_ij_[n];
+				Matd velocity_gradient_ij = -(vel_[index_i] - vel_[index_j]) * nablaW_ijV_j.transpose();
+				velocity_gradient += velocity_gradient_ij;
+			}
+			velocity_gradient_[index_i] = velocity_gradient;
 		}
 		//=================================================================================================//
 		template <class RiemannSolverType>
 		void BaseStressRelaxation1stHalf<RiemannSolverType>::update(size_t index_i, Real dt)
 		{
-			vel_[index_i] += (acc_prior_[index_i] + acc_[index_i]) * dt;
+			Mat3d velocity_gradient = increaseTensor(velocity_gradient_[index_i]);
+			Mat3d stress_tensor_rate_3D_ = plastic_continuum_.ConstitutiveRelation(velocity_gradient, stress_tensor_3D_[index_i]);
+			stress_rate_3D_[index_i] += stress_tensor_rate_3D_;
+			stress_tensor_3D_[index_i] += stress_rate_3D_[index_i] * dt;
+			//For plasticity
+			Mat3d stress_tensor_ = plastic_continuum_.ReturnMapping(stress_tensor_3D_[index_i]);
+			stress_tensor_3D_[index_i] = stress_tensor_;
+			vertical_stress_[index_i] = stress_tensor_3D_[index_i](1, 1);
+			strain_rate_3D_[index_i] = 0.5 * (velocity_gradient + velocity_gradient.transpose());
+			strain_tensor_3D_[index_i] += strain_rate_3D_[index_i] * dt;
+			//calculate elastic strain
+			Mat3d deviatoric_stress = stress_tensor_3D_[index_i] - (1 / 3) * stress_tensor_3D_[index_i].trace() * Mat3d::Identity();
+			Real hydrostatic_pressure = (1 / 3) * stress_tensor_3D_[index_i].trace();
+			elastic_strain_tensor_3D_[index_i] = deviatoric_stress / (2 * plastic_continuum_.getShearModulus(E_, nu_)) + hydrostatic_pressure * Mat3d::Identity() / (9 * plastic_continuum_.getBulkModulus(E_, nu_));
+			Mat3d plastic_strain_tensor_3D = strain_tensor_3D_[index_i] - elastic_strain_tensor_3D_[index_i];
+			acc_deviatoric_plastic_strain_[index_i] = particles_->getDeviatoricPlasticStrain(plastic_strain_tensor_3D);
 		}
 		//=================================================================================================//
 		template <class RiemannSolverType>
-		Vecd BaseStressRelaxation1stHalf<RiemannSolverType>::computeNonConservativeAcceleration(size_t index_i)
+		BaseStressRelaxation2ndHalf<RiemannSolverType>::BaseStressRelaxation2ndHalf(BaseInnerRelation& inner_relation)
```

Comment:
I think that you have reversed the meaning of the first and second half steps, which me why you need a third half step.

Reply:
I add the third half step because there are many intermediate variables, and updating these variables requires following a certain order of priority (Verlet scheme). If there are just 1st and 2nd steps, there are six places that we can update variables with different priorities, considering the initialization, interaction, and update for each relaxation. But I need more than six places to organize the order of updating variables. That's why I add a third half relaxation.

Reply:
could you specify the the order of updating, so that we can have a look if there is a way.

Reply:
This is the current structure:
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/8315f28f-33d7-4448-abc2-20024a5ecf00)
I think `StressRelaxation2ndHalfInner::update()` and `StressRelaxation3ndHalfInner::initialization()` can be combined together. For other parts, I am not sure how to simplify.

Reply:
Usually, we only put particle interaction method in interaction() function only.

Reply:
> I think that the velocity should be updated together with density change rate (they are very similar things). So you you only need to steps.

Thanks. I will try to check this.